Improvement/Fixup on buckets creations

benzekrimaha opened the pull request:
with the migration to typescript the bucket creation has been changed always passing the capabilities as `undefined` , which is now breaking s3utils as we’re not able to retrieve the capabilities in order to update them

Issue: ARSN-492

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-492 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.2.14`


Please check the `Fix Version/s` of ARSN-492, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -479,19 +479,31 @@ class MongoClientInterface {
                 if (!doc) {
                     return cb(errors.NoSuchBucket);
                 }
+                log.info('getBucketAttributes: bucket found', {
```
seems way too complex for logs, logs should only include what we have...
esp. on `getBucketAttributes` (or any metadata getter), i don't think we want to log every time...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -479,19 +479,31 @@ class MongoClientInterface {
                     quotaMax: doc.value.quotaMax?.toString() || '0',
                     capabilities: {
-                        ...doc.value.capabilities,
+                        ...(doc.value.capabilities || {}),
```
what i the point? `...undefined` is exactly the same as `...{}`

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -479,19 +479,31 @@ class MongoClientInterface {
                         },
                     },
                 };
+                log.info('getBucketAttributes: bucket metadata', {
```
debug log, should be removed

francoisferrand commented on package.json:
```diff
@@ -3,7 +3,7 @@
     "node": ">=20"
   },
-  "version": "8.2.13",
+  "version": "8.2.14",
```
not sure we want to release just yet, we got a few more that also need to reach 9.0 very soon.

benzekrimaha replied:
we will actually need to bump, for the s3utils PR

francoisferrand replied:
yes we need to bump, but not on every PR. We currently have 3 PRs open in arsenal, all of which need to be bump'ed : so better to merge as many PRs as possible, and release/bump just once, after the last one of the batch.

benzekrimaha committed:
capabilites value passing instead of undefined in createBucket

In some tests, we actually pass the capabilities value to the createBucket function, overriding it to undefined breaks the behavior.
This commit fixes that by passing the capabilities value to the createBucket function.
Issue: ARSN-492

williamlardier commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -411,7 +411,8 @@ class MongoClientInterface {
                     quotaMax: new Long(newBucketMD.quotaMax || '0'),
-                    capabilities: undefined,
+                    // We're passing the capabilities value here as it's used for testing purposes
+                    capabilities: newBucketMD.capabilities as CapabilitiesMongoDB|| undefined,
```
```suggestion
                    capabilities: newBucketMD.capabilities as CapabilitiesMongoDB || undefined,
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -411,7 +411,8 @@ class MongoClientInterface {
                     quotaMax: new Long(newBucketMD.quotaMax || '0'),
-                    capabilities: undefined,
+                    // We're passing the capabilities value here as it's used for testing purposes
+                    capabilities: newBucketMD.capabilities as CapabilitiesMongoDB || undefined,
```
that is very weird:
* `newBucketMD` is the "serializable" variant of `bucketMD`: if typing is needed, it should be done in `BucketInfo.makeSerializable`
* adding the type in here will only affect this specific function, as `payload` is only used here
* Passing `bucketMD.capabilities` to createBucket() is not just for testing: today, maybe we do not do it "in prod" and only do it in tests, but since it is part of bucketInfo it should be supported, like all other fields... → if a fix is needed, it is not just for tests...

benzekrimaha replied:
Actually like for the quotaMax , here we don't really need the makeSerializable to make capabilities typed the way it's needed to be set in mongodb as it's only used here thus in the payload used to create the bucket

francoisferrand commented on lib/models/Veeam.ts:
```diff
@@ -72,9 +73,9 @@ export class VeeamCapacityInfo {
 
     static toBigInt(capacity: VeeamSOSApiSchema['CapacityInfo']): VeeamSOSApiSchema['CapacityInfo'] {
```
the return type is not correct : it now has only `bigint` only

benzekrimaha replied:
As we also have the lastModified which is not a big int => keeping the VeeamSOSApiSchema['CapacityInfo'] seems fair...

francoisferrand replied:
the point of typing is to allow the compiler to do the work : if we type too broadly, then typing becomes useless, we have no benefit anymore... The whole point of this function is to convert to bigint, so best to specify the actual type...
```suggestion
    static toBigInt(capacity: VeeamSOSApiSchema['CapacityInfo']): { Capacity: BigInt; Available: BigInt; Used: BigInt; LastModified?: string } {
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -107,21 +107,34 @@ export type MongoDBClientInterfaceParameters = {
+            ? Array<NestedOmit<U, K>>
+            : {
+                [P in keyof T as P extends Extract<K, string> ? never : P]:
+                    NestedOmit<T[P], StripKey<K, P>>;
+            };
 
-export type CapabilitiesMongoDB = Capabilities & {
-    VeeamSOSApi?: Omit<VeeamSOSApiSchema, 'CapacityInfo'> & {
-        CapacityInfo?: {
-            Capacity: Long,
-            Available: Long,
-            Used: Long,
-        },
-    },
-}
+type StripKey<K extends PropertyKey, P extends PropertyKey> =
+    K extends `${Exclude<P, symbol>}.${infer R}` ? R : never;
```
should go to a dedicated file...

francoisferrand commented on lib/models/Veeam.ts:
```diff
@@ -26,9 +27,9 @@ export type VeeamSOSApiSchema = {
-        Used: bigint | number,
+        Capacity: bigint | number | Long,
+        Available: bigint | number | Long,
+        Used: bigint | number | Long,
```
why is it needed?
AFAIK we want to convert to/from Long inside MongoClientInterface : so the type handled throughout the code should not have `Long` anywhere...

benzekrimaha replied:
Actually this was done for the VeeamCapacityInfo serialize function here to be able to proceed , the thing is that we're storing this as a long in mongodb then expecting the variable to be typed as `bigint | number` on the function as it's taking the` capacity: VeeamSOSApiSchema['CapacityInfo']` as a param , which will not be the case as we're using the doc.value.x.x which is a long , this is why long was added in order to make the serialization possible. This will not affect the calculous as we're making the variables bigInts later on for calculous

francoisferrand replied:
this however means that "user" of Capacity need (according to the compiler) to handle the `Long` type : which is exactly what we don't want, we want the `Long` type to stay in Mongo...

a better way seems to specify `bigint` here only (as it is what we want the code to deal with)
```suggestion
        Used: bigint,
```

and change the type of VeeamCapacityInfo.serialize() to be more permissive, like this:

```
static serialize(capacity: {
        Capacity: bigint | number | Long,
        Available: bigint | number | Long,
        Used: bigint | number | Long,
        LastModified?: string,
    }): VeeamSOSApiSerializable['CapacityInfo'] {
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -411,7 +424,18 @@ class MongoClientInterface {
                     ...newBucketMD,
                     quotaMax: new Long(newBucketMD.quotaMax || '0'),
-                    capabilities: undefined,
+                    capabilities: newBucketMD.capabilities ? {
+                        ...newBucketMD.capabilities,
+                        VeeamSOSApi: newBucketMD.capabilities.VeeamSOSApi ? {
+                            ...newBucketMD.capabilities.VeeamSOSApi,
+                            CapacityInfo: newBucketMD.capabilities.VeeamSOSApi.CapacityInfo ? {
+                                Capacity: new Long(newBucketMD.capabilities.VeeamSOSApi.CapacityInfo.Capacity || '0'),
+                                Available: new Long(newBucketMD.capabilities.VeeamSOSApi.CapacityInfo.Available || '0'),
+                                Used: new Long(newBucketMD.capabilities.VeeamSOSApi.CapacityInfo.Used || '0'),
+                                LastModified: (newBucketMD.capabilities.VeeamSOSApi.CapacityInfo as any).LastModified,
+                            } : undefined,
+                        } : undefined,
```
instead of ternary operator, prefer `newBucketMD.capabilities.VeeamSOSApi && { .... }` syntax

```suggestion
                    capabilities: newBucketMD.capabilities && {
                        ...newBucketMD.capabilities,
                        VeeamSOSApi: newBucketMD.capabilities.VeeamSOSApi && {
                            ...newBucketMD.capabilities.VeeamSOSApi,
                            CapacityInfo: newBucketMD.capabilities.VeeamSOSApi.CapacityInfo && {
                                Capacity: new Long(newBucketMD.capabilities.VeeamSOSApi.CapacityInfo.Capacity || '0'),
                                Available: new Long(newBucketMD.capabilities.VeeamSOSApi.CapacityInfo.Available || '0'),
                                Used: new Long(newBucketMD.capabilities.VeeamSOSApi.CapacityInfo.Used || '0'),
                                LastModified: (newBucketMD.capabilities.VeeamSOSApi.CapacityInfo as any).LastModified,
                            },
                        },
                    },
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -411,7 +424,18 @@ class MongoClientInterface {
+                                LastModified: (newBucketMD.capabilities.VeeamSOSApi.CapacityInfo as any).LastModified,
+                            } : undefined,
+                        } : undefined,
+                    } : undefined,
```
missing tests for this case

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/ARSN-492` and the
destination branch `development/8.2`.

Please resolve the conflict on **the feature branch** (`improvement/ARSN-492`).

```sh
git fetch && \
git checkout origin/improvement/ARSN-492 && \
git merge origin/development/8.2
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/ARSN-492
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -19,7 +19,7 @@ import errors, { ArsenalError, errorInstances } from '../../../errors';
 import ObjectMD, { ObjectMDData } from '../../../models/ObjectMD';
 import * as jsutil from '../../../jsutil';
-import { ArsenalCallback } from '../../../types';
+import { ArsenalCallback , NestedOmit } from '../../../types';
```
typo:
```suggestion
import { ArsenalCallback, NestedOmit } from '../../../types';
```

francoisferrand reviewed (commented):
Also please update the PR with proper title and description

francoisferrand commented on lib/models/Veeam.ts:
```diff
@@ -72,9 +78,9 @@ export class VeeamCapacityInfo {
     static toBigInt(capacity: VeeamSOSApiSchema['CapacityInfo']): VeeamSOSApiSchema['CapacityInfo'] {
         return {
-            Capacity: BigInt(capacity?.Capacity || 0),
-            Available: BigInt(capacity?.Available || 0),
-            Used: BigInt(capacity?.Used || 0),
+            Capacity: BigInt(capacity?.Capacity.toString() || 0),
+            Available: BigInt(capacity?.Available.toString() || 0),
+            Used: BigInt(capacity?.Used.toString() || 0),
```
why do we _need_ to convert to string?
- if will actually crash if capacity were null/undefined (`capacity?.`)
- according to the type (and typescript compiler), Capacity/Available/Used can now only be bigint already...

We can keep the `BigInt` call for extra caution, but we should avoid the systematic double conversion to and from string...

benzekrimaha commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.2`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.1`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.1`



Please check the status of the associated issue ARSN-492.

Goodbye benzekrimaha.
